services/metadata/src/metadata/persistence/avu.clj

```diff
@@ -98,25 +103,38 @@
       existing-avu
       (add-avus user-id [avu]))))
 
+(defn- add-orphaned-ids-where-clause
+  [query avu-ids-to-keep]
+  (if (empty? avu-ids-to-keep)
+    query
+    (where query {:id [not-in avu-ids-to-keep]})))
+
 (defn remove-orphaned-avus
-  "Removes AVUs for the given target-id that are not in the given set of avu-ids."
-  [target-type target-id avu-ids]
-  (let [query (-> (delete* :avus)
-                  (where {:target_id target-id
-                          :target_type (db/->enum-val target-type)}))]
-    (if (empty? avu-ids)
-      (delete query)
-      (delete (where query {:id [not-in avu-ids]})))))
+  "Removes AVUs for the given target-id that are not in the given set of avu-ids-to-keep."
+  [target-type target-id avu-ids-to-keep]
+  (let [avus-to-remove (-> (select* :avus)
+                           (target-where-clause target-type target-id)
+                           (add-orphaned-ids-where-clause avu-ids-to-keep)
+                           select)]
+    ;; Remove orphaned sub-AVUs of any AVUs to be removed by this request.
+    (doseq [{:keys [id]} avus-to-remove]
+      (remove-orphaned-avus "avu" id nil))
```

Comment:
This recursive call to `metadata.persistence.avus/remove-orphaned-avus` ensures sub-AVUs of any AVUs about to be removed are cleaned up as well.